feat!: provide a compact form of TransactionInput

philipr-za opened the pull request:
Description
---

This PR updates the TransactionInput struct to support containing the duplicated Output data of the output it is spending OR just the hash of the output. This is done in the original TransactionInput as opposed to creating a new struct to allow for easier reuse of all the code that already supports the TransactionInput.

Methods are provided to:
- Create a new compact or full TransactionInput
- Return a compact form of the current TransactionInput
- Provide the output data that is referenced in the TransactionInput

This PR also updates the RPC definitions and all the conversion methods to support both forms of the TransactionInput

Motivation and Context
---

The reason for this change is to reduce the amount of duplicate data stored in the Blockchain db and sent across the wire during Block sync. Currently the input is stored with all the duplicated data from the output it is spending along side the spent output in the database which is a duplication of data. Now the TransactionInput is converted into its compact form with just a reference to the spent output before being stored in the database and before being sent across the wire. When the Input is read from the database or received then the associated output data is retrieved from the local datastore.

TODO: apply the compact form of the TransactionInput to transactions submitted to the base node mempool. This will further reduce data sent from wallets to the base nodes.

How Has This Been Tested?
---
cargo test

delta1 commented on base_layer/core/src/transactions/transaction.rs:
```diff
@@ -430,13 +474,48 @@ impl TransactionInput {
+        }
+    }
+
+    /// Accessor method for the commitment contained in an input
```
duplicated comments

stringhandler reviewed (commented):
Initial impression is that it looks good, but possibly worth testing on igor or a different network for a bit. Maybe we can set a wire mode?

delta1 reviewed (dismissed):
Fantastic work 👏 

i did some system level testing with a new chain on `igor`: 
- 2 base nodes 
- 2 wallets
- Mined with mining node 
- Merge mined with xmrig and proxy 
- Sent and mined transactions between wallets
- Sent and mined one-sided transactions between wallets

philipr-za committed:
feat!: Provide a compact form of TransactionInput

This PR updates the TransactionInput struct to support containing the duplicated Output data of the output it is spending OR just the hash of the output. This is done in the original TransactionInput as opposed to creating a new struct to allow for easier reuse of all the code that already supports the TransactionInput.

Methods are provided to:
- Create a new compact or full TransactionInput
- Return a compact form of the current TransactionInput
- Provide the output data that is referenced in the TransactionInput

This PR also updates the RPC definitions and all the conversion methods to support both forms of the TransactionInput

The reason for this change is to reduce the amount of duplicate data stored in the Blockchain db and sent across the wire during Block sync. Currently the input is stored with all the duplicated data from the output it is spending along side the spent output in the database which is a duplication of data. Now the TransactionInput is converted into its compact form with just a reference to the spent output before being stored in the database and before being sent across the wire. When the Input is read from the database or received then the associated output data is retrieved from the local datastore.

TODO: apply the compact form of the TransactionInput to transactions submitted to the base node mempool. This will further reduce data sent from wallets to the base nodes.

philipr-za committed:
cargo fmt --all
